### PR TITLE
molecule: use user-provided IDs in molecule files

### PR DESCRIPTION
**Pull request by jrgissing**
**Summary**

previously, the atom IDs in per-atom sections of molecule templates were ignored. though documented, this is rather unintuitive behavior. thanks to Ben Jensen for pointing this out

**Related Issue(s)**

another open PR modifying molecule.* files: #2447 

**Author(s)**

JG

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

**Implementation Notes**

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

File: src/molecule.cpp [hunk 1]
```diff
@@ -683,14 +683,15 @@ void Molecule::coords(char *line)
...
-      values.next_int();
-      x[i][0] = values.next_double();
-      x[i][1] = values.next_double();
-      x[i][2] = values.next_double();
-
-      x[i][0] *= sizescale;
-      x[i][1] *= sizescale;
-      x[i][2] *= sizescale;
+      int iatom = values.next_int() - 1;
+      if (iatom >= natoms) error->one(FLERR,"Invalid Coords section in molecule file");
```
Comment (rbberger) [inline src/molecule.cpp:687]: There should be range check for `iatom < 0` too.

File: src/molecule.cpp [hunk 2]
```diff
@@ -718,9 +719,10 @@ void Molecule::types(char *line)
       ValueTokenizer values(line);
       if (values.count() != 2) error->one(FLERR,"Invalid Types section in molecule file");
 
-      values.next_int();
-      type[i] = values.next_int();
-      type[i] += toffset;
+      int iatom = values.next_int() - 1;
+      if (iatom >= natoms) error->one(FLERR,"Invalid Types section in molecule file");
```
Comment (rbberger) [inline src/molecule.cpp:723]: Same here.

File: src/molecule.cpp [hunk 3]
```diff
@@ -748,9 +750,10 @@ void Molecule::molecules(char *line)
       ValueTokenizer values(line);
       if (values.count() != 2) error->one(FLERR,"Invalid Molecules section in molecule file");
 
-      values.next_int();
-      molecule[i] = values.next_int();
-      // molecule[i] += moffset; // placeholder for possible molecule offset
+      int iatom = values.next_int() - 1;
+      if (iatom >= natoms) error->one(FLERR,"Invalid Molecules section in molecule file");
```
Comment (rbberger) [inline src/molecule.cpp:754]: Same here.

File: src/molecule.cpp [hunk 4]
```diff
@@ -808,8 +811,9 @@ void Molecule::charges(char *line)
       ValueTokenizer values(line);
       if ((int)values.count() != 2) error->one(FLERR,"Invalid Charges section in molecule file");
 
-      values.next_int();
-      q[i] = values.next_double();
+      int iatom = values.next_int() - 1;
+      if (iatom >= natoms) error->one(FLERR,"Invalid Charges section in molecule file");
```
Comment (rbberger) [inline src/molecule.cpp:815]: Same here.

File: src/molecule.cpp [hunk 5]
```diff
@@ -831,11 +835,12 @@ void Molecule::diameters(char *line)
       ValueTokenizer values(line);
       if (values.count() != 2) error->one(FLERR,"Invalid Diameters section in molecule file");
 
-      values.next_int();
-      radius[i] = values.next_double();
-      radius[i] *= sizescale;
-      radius[i] *= 0.5;
-      maxradius = MAX(maxradius,radius[i]);
+      int iatom = values.next_int() - 1;
+      if (iatom >= natoms) error->one(FLERR,"Invalid Diameters section in molecule file");
```
Comment (rbberger) [inline src/molecule.cpp:839]: Same here.

File: src/molecule.cpp [hunk 6]
```diff
@@ -860,9 +865,10 @@ void Molecule::masses(char *line)
       ValueTokenizer values(line);
       if (values.count() != 2) error->one(FLERR,"Invalid Masses section in molecule file");
 
-      values.next_int();
-      rmass[i] = values.next_double();
-      rmass[i] *= sizescale*sizescale*sizescale;
+      int iatom = values.next_int() - 1;
+      if (iatom >= natoms) error->one(FLERR,"Invalid Diameters section in molecule file");
```
Comment (rbberger) [inline src/molecule.cpp:869]: Same here.

**Review by rbberger — CHANGES_REQUESTED** [review]

**Review by rbberger — APPROVED** [review]

**Comment by rbberger** [comment]
@sjplimp Please take a look at this one.

**Comment by sjplimp** [comment]
@jrgissing Not clear on what is being changed or why.  In particular, the doc page was clear (to me)
before, but now it is less clear.  Are you saying with this change the N lines in any per-atom section no longer have to be in order, but presuambly they still need to be complete (N different IDs from 1 to N)?  Also, for the Bonds, Angles, etc sections which list 2 or more atom IDs ?  Those IDs will now match the atom ID (if the atoms are not listed in order), otherwise same as before?  If that is correct, it should be documented more clearly.  Also, I'm not sure it is a good idea to allow users to list atom IDs (or diameters, etc) in some order other than 1 to N.  Is there a motivation for that?  Otherwise the code could verify that the IDs are in order from 1 to N, and thus complete, which with this change would require non-trivial work.

**Comment by jrgissing** [comment]
> @jrgissing Not clear on what is being changed or why. 
Now that template atom IDs may be referenced, it was unexpected for them to be ignored (specifically, in the context of bond/react). 

> In particular, the doc page was clear (to me) before, but now it is less clear. Are you saying with this change the N lines in any per-atom section no longer have to be in order, but presuambly they still need to be complete (N different IDs from 1 to N)? 
Yes.

> Also, for the Bonds, Angles, etc sections which list 2 or more atom IDs ? Those IDs will now match the atom ID (if the atoms are not listed in order), otherwise same as before?
Yes.

> If that is correct, it should be documented more clearly. Also, I'm not sure it is a good idea to allow users to list atom IDs (or diameters, etc) in some order other than 1 to N. Is there a motivation for that? Otherwise the code could verify that the IDs are in order from 1 to N, and thus complete, which with this change would require non-trivial work.
I think the out-of-order issue comes up when molecule templates are created from data files, which don't always keep atom order. I can add a completeness check.

**Comment by jrgissing** [comment]
ok, the completeness check has been added. also eliminates some confusing variable shadowing

**Comment by sjplimp** [comment]
@jrgissing ok - your comments above and code changes look fine.  My only last comment is that the change to the doc page molecule.rst now seems incorrect?  Should it say the lines in per-atom sections need to include a setting for every atom (IDs for 1 to Natoms) but the lines can be in any order?

**Comment by rbberger** [comment]
@sjplimp please approve for merge

**Review by sjplimp — APPROVED** [review]

**Pull request merged**